Sources/BraintreeVenmo/BTVenmoClient.swift

```diff
@@ -0,0 +1,357 @@
+
+            if let displayName = request.displayName {
+                inputParameters["displayName"] = displayName
+            }
+
+            let inputDictionary: [String: Any] = ["input": inputParameters]
+            let graphQLParameters: [String: Any] = [
+                "query": "mutation CreateVenmoPaymentContext($input: CreateVenmoPaymentContextInput!) { createVenmoPaymentContext(input: $input) { venmoPaymentContext { id } } }",
+                "variables": inputDictionary
+            ]
+
+            self.apiClient.post("", parameters: graphQLParameters, httpType: .graphQLAPI) { body, _, error in
+                if let error = error as? NSError {
+                    if error.code == BTCoreConstants.networkConnectionLostCode {
+                        self.apiClient.sendAnalyticsEvent("ios.pay-with-venmo.network-connection.failure")
+                    }
+
+                    completion(nil, BTVenmoError.invalidRedirectURL("Failed to fetch a Venmo paymentContextID while constructing the requestURL."))
+                    return
+                }
+
+                guard let body else {
+                    completion(nil, BTVenmoError.invalidBodyReturned)
+                    return
+                }
+
+                guard let paymentContextID = body["data"]["createVenmoPaymentContext"]["venmoPaymentContext"]["id"].asString() else {
+                    completion(nil, BTVenmoError.invalidRedirectURL("Failed to parse a Venmo paymentContextID while constructing the requestURL. Please contact support."))
+                    return
+                }
+
+                guard let appSwitchURL = BTVenmoAppSwitchRedirectURL().appSwitch(
+                    returnURLScheme: returnURLScheme,
+                    forMerchantID: merchantProfileID,
+                    accessToken: configuration.venmoAccessToken,
+                    bundleDisplayName: bundleDisplayName,
+                    environment: configuration.venmoEnvironment,
+                    paymentContextID: paymentContextID,
+                    metadata: metadata
+                ) else {
+                    completion(nil, BTVenmoError.invalidRedirectURL("The request URL could not be constructed or was nil."))
+                    return
+                }
+
+                self.performAppSwitch(with: appSwitchURL, shouldVault: request.vault, completion: completion)
+            }
+        }
+    }
+
+    /// Returns true if the proper Venmo app is installed and configured correctly, returns false otherwise.
+    public func isVenmoAppInstalled() -> Bool {
+        if let _ = application as? UIApplication {
+            guard let appSwitchURL = BTVenmoAppSwitchRedirectURL().baseAppSwitchURL else {
+                return false
+            }
+
+            return UIApplication.shared.canOpenURL(appSwitchURL)
+        } else {
+            return application.canOpenURL(BTVenmoAppSwitchRedirectURL().baseAppSwitchURL ?? URL(string: "")!)
+        }
```

Comment:
This is hard to read, IMO - as well as the way we inject the `application` mock. 

I think mocking `UIApplication` is a good case for using `OCMock`, and re-writing[ `FakeApplication`](https://github.com/braintree/braintree_ios/blob/master/UnitTests/BraintreeTestShared/FakeApplication.swift) in ObjC. This would allow us to remove the `AnyObject` type on line 20 we have now to work around Swift's mocking limitations.

Eventually, this would allow us to remove testing details from the source code:
```
        guard let appSwitchURL = BTVenmoAppSwitchRedirectURL().baseAppSwitchURL else {
            return false
        }
        
        return application.canOpenURL(appSwitchURL)
```

Definitely a punt 🏈 but wanted to leave a note. One example of this is in [BTMockOpenURLContext](https://github.com/braintree/braintree_ios/blob/master/UnitTests/BraintreeCoreTests/Helpers/BTMockOpenURLContext.h).

Reply:
It appears that the conversation is related to improving the readability and maintainability of the codebase, specifically in relation to mocking the UIApplication and the BTVenmoAppSwitchRedirectURL. The proposed solution involves using OCMock and rewriting the FakeApplication class in Objective-C to remove Swift's mocking limitations, which would allow for the removal of testing details from the source code. This suggestion has been noted for a follow-up PR, with BTMockOpenURLContext given as an example.

Reply:
Totally agree - this is much harder to reason about in Swift. It wasn't great in Obj-C either, but was obfuscated by the... Objective-C-ness. 

It looks like there could be some ways around this in Swift. I found [this blog post](http://swiftyjimmy.com/mock-uiapplication-swift/) but may also be worth a punt to explore if there is a better way to do this so we can avoid these weird switches in the implementation.

Reply:
Ah yeh, that's like what we're doing on PPCP for `URLSession`. Still a pain to need to make those protocols, but such is life! Swift can't be better at _everything_ :D 

Ticket for improving this later - DTBTSDK-2517